token/token.go

```diff
@@ -1,75 +1,146 @@
-func NewManagerWithPrivateKey(privateKey *rsa.PrivateKey) Manager {
-	return &tokenManager{
-		publicKey:  &privateKey.PublicKey,
-		privateKey: privateKey,
-	}
-}
-
-func (mgm tokenManager) Extract(tokenString string) (*account.Identity, error) {
-	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
-		return mgm.publicKey, nil
-	})
+	remoteKeys, err := authtoken.FetchKeys(config.GetKeysEndpoint())
 	if err != nil {
-		return nil, errors.WithStack(err)
+		log.Error(nil, map[string]interface{}{
+			"keys_url": config.GetKeysEndpoint(),
+		}, "unable to load public keys from remote service")
+		return nil, errors.New("unable to load public keys from remote service")
+	}
+	for _, remoteKey := range remoteKeys {
+		tm.publicKeysMap[remoteKey.KeyID] = remoteKey.Key
+		tm.publicKeys = append(tm.publicKeys, &PublicKey{KeyID: remoteKey.KeyID, Key: remoteKey.Key})
+		log.Info(nil, map[string]interface{}{
+			"kid": remoteKey.KeyID,
+		}, "Public key added")
 	}
 
-	if !token.Valid {
-		return nil, errors.New("Token not valid")
+	devModeURL := config.GetKeycloakDevModeURL()
+	if devModeURL != "" {
+		remoteKeys, err = authtoken.FetchKeys(fmt.Sprintf("%s/protocol/openid-connect/certs", devModeURL))
+		if err != nil {
+			log.Error(nil, map[string]interface{}{
+				"keys_url": devModeURL,
+			}, "unable to load public keys from remote service in Dev Mode")
+			return nil, errors.New("unable to load public keys from remote service  in Dev Mode")
+		}
+		for _, remoteKey := range remoteKeys {
+			tm.publicKeysMap[remoteKey.KeyID] = remoteKey.Key
+			tm.publicKeys = append(tm.publicKeys, &PublicKey{KeyID: remoteKey.KeyID, Key: remoteKey.Key})
+			log.Info(nil, map[string]interface{}{
+				"kid": remoteKey.KeyID,
+			}, "Public key added")
+		}
 	}
 
-	claimedUUID := token.Claims.(jwt.MapClaims)["sub"]
-	if claimedUUID == nil {
-		return nil, errors.New("Subject can not be nil")
+	return tm, nil
+}
+
+// NewManagerWithPublicKey returns a new token Manager for handling tokens with the only public key
+func NewManagerWithPublicKey(id string, key *rsa.PublicKey) Manager {
+	return &tokenManager{
+		publicKeysMap: map[string]*rsa.PublicKey{id: key},
+		publicKeys:    []*PublicKey{{KeyID: id, Key: key}},
 	}
-	// in case of nil UUID, below type casting will fail hence we need above check
-	id, err := uuid.FromString(token.Claims.(jwt.MapClaims)["sub"].(string))
+}
+
+// ParseToken parses token claims
+func (mgm *tokenManager) ParseToken(ctx context.Context, tokenString string) (*TokenClaims, error) {
+	token, err := jwt.ParseWithClaims(tokenString, &TokenClaims{}, func(token *jwt.Token) (interface{}, error) {
+		kid := token.Header["kid"]
+		if kid == nil {
+			log.Error(ctx, map[string]interface{}{}, "There is no 'kid' header in the token")
+			return nil, errors.New("There is no 'kid' header in the token")
+		}
+		key := mgm.PublicKey(fmt.Sprintf("%s", kid))
+		if key == nil {
+			log.Error(ctx, map[string]interface{}{
```

Comment:
same comment as above: what about wrapping the error with the new message ?

Reply:
To log the key ID parameter.